crates/hc_sandbox/src/zome_call.rs

```diff
@@ -0,0 +1,516 @@
+
+#[derive(Debug, Parser)]
+pub struct ConnectArgs {
+    /// Ports to running conductor admin interfaces.
+    #[arg(short, long, conflicts_with_all = &["indices"], value_delimiter = ',')]
+    pub running: Option<u16>,
+
+    /// Select from existing conductor sandboxes specified in `$(pwd)/.hc` by index.
+    #[arg(short, long, conflicts_with_all = &["running"])]
+    pub index: Option<u32>,
+}
+
+/// Create authentication credentials for making zome calls and deploy them to Holochain.
+#[derive(Debug, Parser)]
+pub struct ZomeCallAuth {
+    #[command(flatten)]
+    pub connect_args: ConnectArgs,
+
+    /// Whether to pipe the passphrase from stdin.
+    ///
+    /// By default, the passphrase is read interactively from the user.
+    #[arg(long)]
+    pub piped: bool,
+
+    /// The installed app id to authorize calls for.
+    pub app_id: String,
+}
+
+/// Make a zome call to an app on a running conductor.
+#[derive(Debug, Parser)]
+pub struct ZomeCall {
+    #[command(flatten)]
+    pub connect_args: ConnectArgs,
+
+    /// Whether to pipe the passphrase from stdin.
+    ///
+    /// By default, the passphrase is read interactively from the user.
+    #[arg(long)]
+    pub piped: bool,
+
+    /// The installed app id to call a function for
+    pub app_id: String,
+
+    /// The DNA hash to call
+    pub dna_hash: DnaHashB64,
+
+    /// The zome to call
+    pub zome_name: String,
+
+    /// The zome function to call
+    pub function: String,
+
+    /// The zome call payload as JSON
+    pub payload: String,
+}
+
+pub async fn zome_call_auth(
+    zome_call_auth: ZomeCallAuth,
+    admin_port: Option<u16>,
+) -> anyhow::Result<()> {
+    let admin_port = admin_port_from_connect_args(zome_call_auth.connect_args, admin_port).await?;
+
+    let app_client = AppClient::try_new(admin_port, zome_call_auth.app_id.clone()).await?;
+    let app_info = app_client.request(AppRequest::AppInfo).await?;
+    let info = match app_info {
+        AppResponse::AppInfo(Some(info)) => info,
+        other => anyhow::bail!("Unexpected response while getting app info: {:?}", other),
+    };
+
+    let cell_ids = info
+        .cell_info
+        .values()
+        .flatten()
+        .filter_map(|info| match info {
+            CellInfo::Provisioned(info) => Some(info.cell_id.clone()),
+            _ => None,
+        })
+        .collect::<HashSet<_>>();
+
+    let mut client = CmdRunner::try_new(admin_port).await?;
+
+    holochain_util::pw::pw_set_piped(zome_call_auth.piped);
+    if !zome_call_auth.piped {
+        msg!("Enter new passphrase to authorize zome calls: ");
+    }
+    let passphrase = holochain_util::pw::pw_get().context("Failed to get passphrase")?;
+
+    let (auth, key) = generate_signing_credentials(passphrase)?;
+
+    let signing_agent_key = AgentPubKey::from_raw_32(key.verifying_key().as_bytes().to_vec());
+
+    for cell_id in cell_ids {
+        client
+            .command(AdminRequest::GrantZomeCallCapability(Box::new(
```

Comment:
I really dislike that we use this mechanism for remote calls and for local calls with no distinction. For one, it's not very reasonable to say that credentials should be accepted regardless of the source. We should be able to specify whether we're accepting calls over the app interface locally, from remote calls, or either. It should just be either by default.

It also seems really silly to create remote call credentials for myself that have to be specified per cell. If I'm working locally, I clearly want to be able to access all my cells. Sharing an auth mechanism with remote calls where granular access is actually wanted seems wrong to me. I think there should be an entirely separate mechanism for app interface calls that is simpler to configure.

Reply:
I mean to remember that this cell level authorizing credential issue was also something that was making it hard to find a solution to securely authorize an external client like a browser tab via UX in launcher (because the browser tab does not have access to the admin websocket) in a way that it would keep working if new cloned cells get created.

Reply:
we never implemented a solution in the end but that was one of the blocker I think

Reply:
Yep, I do think this needs a solution. It forces you to either have a questionable experience where some apps will behave properly if they have all their cells up front and others will need some fiddly solution to authorizing access to new cells. Or you do what launcher has done and sign using the agent keypair which I think is not ideal because it means you need a setup where the UI is close to Lair. Which e.g. prevents you running Holochain on a home server and running the UI on a webserver that you access from another machine.

Reply:
Opened an issue https://github.com/holochain/holochain/issues/4595